tools/pipesys/src/cmd/link.rs

```diff
@@ -0,0 +1,194 @@
+/// a relative path, assume the current process directory is intended.
+fn parent_dir(path: &Path) -> Result<PathBuf> {
+    if path.is_absolute() {
+        if let Some(parent) = path.parent() {
+            return Ok(parent.to_path_buf());
+        }
+    } else if let Ok(current) = env::current_dir() {
+        return Ok(current);
+    }
+    bail!("failed to find parent directory for {}", path.display());
+}
+
+/// Helper function to create stdout and stderr streams with the same file storage.
+async fn output_streams(path: &Path) -> Result<(std::fs::File, std::fs::File)> {
+    let stdout = fs::File::create(path).await?;
+    let stderr = stdout.try_clone().await?;
+    Ok((stdout.into_std().await, stderr.into_std().await))
+}
+
+/// Returns true if a symlink exists at the path, and false otherwise.
+async fn symlink_found(path: &Path) -> bool {
+    let res = fs::symlink_metadata(path)
+        .await
+        .with_context(|| format!("failed to query metadata for {}", path.display()));
+
+    if let Ok(m) = res {
+        if m.is_symlink() {
+            trace!("found symlink for {}", path.display());
+            return true;
+        }
+    }
+
+    trace!("no symlink found for {}", path.display());
+    false
+}
+
+/// Returns false if a symlink exists at the path, and true otherwise.
+async fn symlink_not_found(path: &Path) -> bool {
+    !symlink_found(path).await
+}
+
+/// Initialize an inotify instance with the requested watch mask.
+fn inotify_init(path: &Path, watch_mask: WatchMask) -> Result<Inotify> {
+    let parent_dir = parent_dir(path)?;
+
+    let inotify = Inotify::init().with_context(|| "failed to initialize inotify")?;
+    inotify
+        .watches()
+        .add(&parent_dir, watch_mask)
+        .with_context(|| format!("failed to add watch for {}", parent_dir.display()))?;
+
+    Ok(inotify)
+}
+
+/// Wait for inotify events until the provided function returns true.
+async fn inotify_wait<'a, F>(
+    watcher: Inotify,
+    path: &'a Path,
+    check: &dyn Fn(&'a Path) -> F,
+) -> Result<()>
+where
+    F: Future<Output = bool> + 'a,
+{
+    // Check whether the provided function succeeds right away, in case the watched file was
+    // created or removed before the inotify watch could be started.
+    if check(path).await {
+        return Ok(());
+    }
+
+    let file_name = path
+        .file_name()
+        .with_context(|| format!("failed to find file name for {}", path.display()))?;
+
+    let mut buf = [0; 256];
+    let mut event_stream = watcher
+        .into_event_stream(&mut buf)
+        .with_context(|| "failed to get event stream")?;
+
+    while let Some(event) = event_stream.next().await {
+        let event = event.with_context(|| "failed to read event".to_string())?;
+        if let Some(event_file_name) = event.name {
+            if event_file_name == file_name {
+                trace!("found event for {}", path.display());
+                if check(path).await {
```

Comment:
Is this `fs`?

```suggestion
                if fs::check(path).await {
```

Reply:
> Is this `fs`?

`check` is a function parameter - maybe it needs a better name?